visier/connector/sessions.py

```diff
@@ -218,31 +220,43 @@ def _connect_auth_code(self, auth: OAuth2):
         """Connect to Visier using (three-legged) OAuth2.
         This method will attempt to open a browser for the authentication and consent screens.
         It will also spin up a local web server to receive the OAuth2 authorization code."""
-        url_prefix = auth.host + "/v1/auth/oauth2"
+        def mk_pkce_pair():
+            code_verifier = secrets.token_urlsafe(64)
+            code_challenge_digest = hashlib.sha256(code_verifier.encode()).digest()
+            code_challenge = base64.urlsafe_b64encode(code_challenge_digest).decode().rstrip("=")
+            return code_verifier, code_challenge
 
-        def get_token(code: str) -> str:
+        def get_token(code: str, code_verifier: str) -> str:
             body = {
                 "grant_type": "authorization_code",
                 "client_id": auth.client_id,
                 "scope": "read",
-                "code": code
+                "code": code,
+                "code_verifier": code_verifier
             }
             return self._request_token(auth, body)
 
-        callback_url = os.getenv(ENV_VISIER_OAUTH_CALLBACK_URL)
-        binding = CallbackBinding(callback_url)
+        code_verifier, code_challenge = mk_pkce_pair()
+        url_prefix = auth.host + "/v1/auth/oauth2"
+        binding = CallbackBinding(auth.redirect_uri)
         with CallbackServer(binding) as svr:
+            query_args = {
+                "apikey": auth.api_key,
+                "response_type": "code",
+                "client_id": auth.client_id,
+                "code_challenge_method": "S256",
+                "code_challenge": code_challenge
+            }
             if auth.redirect_uri:
-                redirect_uri_arg = f"&redirect_uri={urllib.parse.quote(auth.redirect_uri, safe='')}"
-            else:
-                redirect_uri_arg = "" # Empty means use the redirect_uri registered with Visier
-            browser_url = f'{url_prefix}/authorize?apikey={auth.api_key}&response_type=code&client_id={auth.client_id}{redirect_uri_arg}'
+                query_args["redirect_uri"] = auth.redirect_uri
+
+            browser_url = f'{url_prefix}/authorize?{urlencode(query_args)}'
             # Launch the browser for authentication and consent screens
             webbrowser.open(browser_url)
             try:
                 # Wait up to 2 minutes for the user to complete the OAuth2 code flow
                 code = svr.queue.get(block=True, timeout=120)
```

Comment:
Just curious why this timeout is different than the `self. _timeout` specified earlier; if it's just arbitrary then that's fine too.

Reply:
The `_timeout` is a per request timeout. The 2min timeout is for the end-to-end authorization code flow. I've refactored this slightly to make that part clearer.